lua/ui/game/pause.lua

```diff
@@ -0,0 +1,112 @@
+--** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+--** OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+--** SOFTWARE.
+--******************************************************************************************************
+
+---@type integer
+local OnPauseClientIndex = -1
+
+---@type number
+local OnPauseTimestamp = 0
+
+---@type number
+local PauseThreshold = 10
+
+---@return integer  # The index of the client, like the parameter `pausedBy` of OnPause
+---@return Client?  # The data of the client
+local function FindLocalClient()
+    local allClients = GetSessionClients()
+    for k = 1, table.getn(allClients) do
+        local client = allClients[k]
+        if client["local"] then
+            return k, client
+        end
+    end
+
+    return -1, nil
+end
+
+--- Called by the engine when the simulation pauses for all clients.
+---@param pausedBy integer  # The index of the client in the clients list (that you get via `GetSessionClients`)
+---@param timeoutsRemaining number
+function OnPause(pausedBy, timeoutsRemaining)
+    -- keep track of who paused and when
+    OnPauseClientIndex = pausedBy
+    OnPauseTimestamp = GetSystemTimeSeconds()
+end
+
+--- Called by the engine when the simulation resumed for all clients.
+function OnResume()
+    OnPauseClientIndex = -1
+    OnPauseTimestamp = 0
+end
+
+-- Called immediately by the engine on the machine that initiated the pause. This function is called only by the client that is initiating the (un)pause.
+function OnUserPause(pause)
+end
+
+local oldSessionRequestPause = _G.SessionRequestPause
+_G.SessionRequestPause = function()
+    -- makes no sense to request a pause on top of a pause
+    if SessionIsPaused() then
+        return
+    end
+
+    oldSessionRequestPause()
+end
+
+local oldSessionResume = _G.SessionResume
+---@return 'Accepted' | 'Declined'
+_G.SessionResume = function()
+    LOG("SessionResume")
+
+    -- scenario's that are all good
+    if SessionIsReplay() or
+        not SessionIsMultiplayer()
+    then
+        oldSessionResume()
+        return 'Accepted'
+    end
+
+    -- local client initiated the pause, we're all good
+    local localClientIndex, clientData = FindLocalClient()
+    if OnPauseClientIndex == localClientIndex then
+        LOG("Same client unpauses")
+        oldSessionResume()
+        return 'Accepted'
+    end
+
+    -- we waited long enough, we're good
+    local timeDifference = GetSystemTimeSeconds() - OnPauseTimestamp
+    if timeDifference > PauseThreshold then
+        LOG("Threshold is met to resume")
+        oldSessionResume()
+        return 'Accepted'
+    else
+        -- inform other clients
+        SessionSendChatMessage(import('/lua/ui/game/clientutils.lua').GetAll(), {
+            to = 'all',
+            text = string.format('Wants to resume the game but has to wait %d seconds', PauseThreshold - timeDifference),
```

Comment:
Should have LOC tags for text. The time should also be rounded up so you can't tell the player to wait 0 seconds.
```suggestion
            text = LOCF('<LOC pause_0003>Wants to resume the game but has to wait %d seconds', math.ceil(PauseThreshold - timeDifference)),
```
We already have:
```
pause_0000="By %s"
pause_0001="Game Resumed"
pause_0002="Game Paused"
```

Reply:
Chat messages (in the current setup) are not localized. This suggestion would fall in the category of properly supporting system messages, which is what @4z0t is suggesting below too.

I think it is outside of the scope of this pull request. If we would like to support system messages we first would need to do some refactoring in my point of view.